### PR TITLE
	Enable navigate-to searching in parallel across a single project.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Note: this is similar to the approach we take with FindRefs.  We process projects serially, but process all the documents in a project in parallel. This ensures that we're not absolutely hammering the system by loading and processing tons of compilations at the same time.  Instead, we have at most one compilation at a time that we're looking at.  

This will likely mean more churn while processing that single project.  However, as we are OOP, we can greatly benefit from the throughput gains, while not having the same concern about the latency impact caused by a GC.

File: src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+                var documentSearchTasks = ArrayBuilder<Task>.GetInstance();
+                var tempArrays = ArrayBuilder<ArrayBuilder<INavigateToSearchResult>>.GetInstance();
+
+                try
+                {
+                    // Spawn off a task per document to search.
+                    SpawnDocumentSearchTasks(documentSearchTasks, tempArrays);
+
+                    // Wait for all the tasks to finish.
+                    await Task.WhenAll(documentSearchTasks).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs]: Nope.  WhenAll takes no cancellation.

File: src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+                }
+            }
+
+            private void SpawnDocumentSearchTasks(ArrayBuilder<Task> tasks, ArrayBuilder<ArrayBuilder<INavigateToSearchResult>> tempArrays)
+            {
+                foreach (var document in _project.Documents)
+                {
+                    _cancellationToken.ThrowIfCancellationRequested();
+
+                    if (_searchDocument != null && document != _searchDocument)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs]: I can't.  I'm comparing it to each individual document in _project.Documents.

File: src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+                    {
+                        continue;
+                    }
+
+                    var tempArray = ArrayBuilder<INavigateToSearchResult>.GetInstance();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs]: sure.

File: src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+                    tempArrays.Add(tempArray);
+
+                    // Kick off a fresh task so searching can happen in parallel.
+                    tasks.Add(Task.Run(() => SearchDocumentAsync(document, tempArray)));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs]: yes.  that's totally wrong.  will fix.  ack!

File: src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+                }
+            }
+
+            private async Task SearchDocumentAsync(
+                Document document, ArrayBuilder<INavigateToSearchResult> results)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs]: sure.

**Comment by CyrusNajmabadi** [comment]
retest this please

File: src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs (hunk 5 above)
Comment (sharwell) [inline src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs]: :bulb: If you pass the cancellation token to `Task.Run` it's extremely unlikely that you'll hit this line with cancellation requested. The aggregate penalty for simply removing this line will likely be smaller than is measurable.

File: src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+                // Delay creating a semantic model until necessary.  But once we create it,
+                // cache for the remainder of the search so that it doesn't get cleaned up.
+                SemanticModel semanticModel = null;
+
+                var declarationInfo = await document.GetSyntaxTreeIndexAsync(_cancellationToken).ConfigureAwait(false);
+                foreach (var declaredSymbolInfo in declarationInfo.DeclaredSymbolInfos)
+                {
+                    if (_cancellationToken.IsCancellationRequested)
+                    {
+                        continue;
```
Comment (sharwell) [inline src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs]: :question: Why not return?

:question: Why return partial results for this cancellation, but in other cases we transition the task to a cancelled state?

File: src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs]: That's probably true.  But it means that this code needs to know that it is being called directly from Task.Run.  I prefer methods not have to care and to generally follow the rules of:

1. At the entry point check for cancellation.
2. During loops check for cancellation.

This ensures that code doesn't really have to care about interprocedural relationships, and it means we can refactor things and not have to think about what cancellation checks we should or should not have.

File: src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs]: Yeah.  I don't know what's up with that.  Will change.

**Comment by Pilchie** [comment]
@CyrusNajmabadi did some form of this get merged already?

**Pull request closed without merging**